### PR TITLE
SONAR-9105 add ws projects change visibility

### PR DESCRIPTION
**Pull request by sns-seb**
🌊  This PR is based on #1932, which should be reviewed first.

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 1]
```diff
@@ -123,6 +125,8 @@
   /**
    * Ensures that {@link #hasComponentUuidPermission(String, String)} is {@code true},
    * otherwise throws a {@link org.sonar.server.exceptions.ForbiddenException}.
+   *
+   * @deprecated use {@link #checkComponentPermission(String, ComponentDto)} instead
    */
   UserSession checkComponentUuidPermission(String permission, String componentUuid);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: MAJOR Add the missing @Deprecated annotation. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java [hunk 2]
```diff
@@ -111,6 +111,8 @@
    * Using {@link #hasComponentPermission(String, ComponentDto)} is recommended
    * because it does not have to load project if the referenced component
    * is not a project.
+   *
+   * @deprecated use {@link #hasComponentPermission(String, ComponentDto)} instead
    */
   boolean hasComponentUuidPermission(String permission, String componentUuid);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/user/UserSession.java]: MAJOR Add the missing @Deprecated annotation. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AuthorizationDao.java#L40](https://github.com/SonarSource/sonarqube/blob/1989411cb9e1b1199132dee73081133d4759a806/server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java#L40): Remove this unused "USER_ID_PARAM" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AuthorizationDao.java#L40](https://github.com/SonarSource/sonarqube/blob/1a4a3c078b0d67060afb132ce5aa2b46e8867713/server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java#L40): Remove this unused "USER_ID_PARAM" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AuthorizationDao.java#L40](https://github.com/SonarSource/sonarqube/blob/52de285ddcfacc6ed814b53c626254f57414b961/server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java#L40): Remove this unused "USER_ID_PARAM" private field. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1068)

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import org.sonar.server.user.UserSession;
+
+import static java.util.Collections.singletonList;
+import static org.sonar.core.permission.ProjectPermissions.PUBLIC_PERMISSIONS;
+import static org.sonar.server.ws.KeyExamples.KEY_PROJECT_EXAMPLE_001;
+import static org.sonar.server.ws.WsUtils.checkRequest;
+import static org.sonarqube.ws.client.project.ProjectsWsParameters.PARAM_PROJECT;
+
+public class ChangeVisibilityAction implements ProjectsWsAction {
+  private static final String ACTION = "change_visibility";
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: We have an inconsistency in the api urls: `api/organizations/update_project_visibility` vs. `api/projects/change_visibility`. I would suggest `api/projects/update_visibility`, but we can also wait for lambda to either complain about it or not.

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn();
+
+    String projectKey = request.mandatoryParam(PARAM_PROJECT);
+    boolean changeToPrivate = PRIVATE_VISIBILITY.equals(request.mandatoryParam(PARAM_VISIBILITY));
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ComponentDto component = componentFinder.getByKey(dbSession, projectKey);
+      checkRequest(isRoot(component), "Component must either be a project or a view");
+      checkRequest(!changeToPrivate || !Qualifiers.VIEW.equals(component.qualifier()), "Views can't be made private");
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: If views are always public, they should either not be allowed for this ws at all, or it should at least be part of the api doc.

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      userSession.checkComponentPermission(UserRole.ADMIN, component);
+      checkRequest(noPendingTask(dbSession, component), "Component visibility can't be changed as long as it has background task(s) pending or in progress");
+
+      if (changeToPrivate != component.isPrivate()) {
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: Let's move this check up for not to fail, if there is nothing to be done anyways.

I would suggest: if no change has to be done, ignore pending tasks, missing permissions and the "Views can't be made private" check.

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java [hunk 6]
```diff
@@ -119,7 +119,7 @@ private boolean noPendingTask(DbSession dbSession, ComponentDto rootComponent) {
   private void updatePermissionsToPrivate(DbSession dbSession, ComponentDto component) {
     // delete project permissions for group AnyOne
     dbClient.groupPermissionDao().deleteByRootComponentIdAndGroupId(dbSession, component.getId(), null);
-    // grant UserRole.CODEVIEWER and UserRole.USER
+    // grant UserRole.CODEVIEWER and UserRole.USER to any group or user with at least one permission on project
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: (Could be part of the previous commit.)

File: sonar-ws/src/main/java/org/sonarqube/ws/client/project/ChangeVisibilityRequest.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.ws.client.project;
+
+public class ChangeVisibilityRequest {
+  private final String project;
+  private final String visibility;
+
+  public enum Visibility {
+    PUBLIC, PRIVATE
```
Comment (bartfastiel) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/project/ChangeVisibilityRequest.java]: I would prefer to make the api-value part of this enum definition, like that:

```PUBLIC("public"), PRIVATE("private")```

File: sonar-ws/src/main/java/org/sonarqube/ws/client/project/ProjectsService.java [hunk 8]
```diff
@@ -106,4 +106,12 @@ public SearchWsResponse search(SearchWsRequest request) {
       .setParam(PAGE_SIZE, request.getPageSize());
     return call(get, SearchWsResponse.parser());
   }
+
+  public void changeVisibility(ChangeVisibilityRequest request) {
+    PostRequest post = new PostRequest(path("change_visibility"))
+        .setParam(PARAM_PROJECT, request.getProject())
+        .setParam("visibility", request.getVisibility());
```
Comment (bartfastiel) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/project/ProjectsService.java]: I would prefer, that `getVisibility()` returns the enum constant, and we extract the string value here by calling something like `getVisibility().getApiValue()`.

File: it/it-tests/src/test/java/it/authorisation/IssuePermissionTest.java [hunk 9]
```diff
@@ -57,21 +57,23 @@ public void init() {
...
+    orchestrator.getServer().provisionProject("privateProject", "PrivateProject");
+    ItUtils.newAdminWsClient(orchestrator).projects().changeVisibility(new ChangeVisibilityRequest("privateProject", PRIVATE));
+    orchestrator.getServer().associateProjectToQualityProfile("privateProject", "xoo", "one-issue-per-line");
+    SonarScanner privateProject = SonarScanner.create(projectDir("shared/xoo-sample"))
+      .setProperty("sonar.projectKey", "privateProject")
+      .setProperty("sonar.projectName", "PrivateProject");
+    orchestrator.executeBuild(privateProject);
+
+    orchestrator.getServer().provisionProject("publicProject", "PublicProject");
+    orchestrator.getServer().associateProjectToQualityProfile("publicProject", "xoo", "one-issue-per-line");
```
Comment (bartfastiel) [inline it/it-tests/src/test/java/it/authorisation/IssuePermissionTest.java]: We _could_ set the visibility explicitly to public. (Nice to have)

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/GroupPermissionMapper.java [hunk 10]
```diff
@@ -51,6 +52,13 @@ void delete(@Param("permission") String permission, @Param("organizationUuid") S
   void selectAllPermissionsByGroupId(@Param("organizationUuid") String organizationUuid,
     @Param("groupId") Integer groupId, ResultHandler resultHandler);
 
+  /**
+   * Lists id of groups with at least one permission on the specified root component but which do not have the specified
+   * permission, <strong>excluding group "AnyOne"</strong> (which implies the returned {@code Sett} can't contain
+   * {@code null}).
+   */
+  Set<Integer> selectGroupIdsWithPermissionOnProjectBut(@Param("projectId") long projectId, @Param("role") String permission);
```
Comment (sns-seb) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/GroupPermissionMapper.java]: yes, it's mandatory, we don't want to retrieve all groups with permission on a given project, it's an unlimited number

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/GroupPermissionDao.java [hunk 11]
```diff
@@ -22,6 +22,7 @@
 import java.util.HashMap;
```
Comment (sns-seb) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/GroupPermissionDao.java]: I'll mention mention of the class

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java (hunk 3 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: I agree on fixing change to update for consistency

File: server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java (hunk 4 above)
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/project/ws/ChangeVisibilityAction.java]: afaik, this is temporary. View may very well be made private too

File: sonar-ws/src/main/java/org/sonarqube/ws/client/project/ChangeVisibilityRequest.java (hunk 7 above)
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/project/ChangeVisibilityRequest.java]: the fact the enum value is not used as such (ie. in upper case) is implementation detail which do not need to be exposed in a public enum

I've made getters default protected for consistency

File: sonar-ws/src/main/java/org/sonarqube/ws/client/project/ProjectsService.java (hunk 8 above)
Comment (sns-seb) [inline sonar-ws/src/main/java/org/sonarqube/ws/client/project/ProjectsService.java]: I'll use the enum in the `ChangeVisibilityRequest` class but won't update the enum to not expose implementation details

**Pull request merged**